torch/optim/optimizer.py

```diff
@@ -236,6 +236,51 @@ class Optimizer:
     _optimizer_load_state_dict_pre_hooks: 'OrderedDict[int, Callable[["Optimizer", StateDict], Optional[StateDict]]]'
     _optimizer_load_state_dict_post_hooks: 'OrderedDict[int, Callable[["Optimizer"], None]]'
 
+
+    # The following hook is automatically registered to load_state_dict as a pre-hook. This processing
+    # used to live within load_state_dict, but, since the introduction of state_dict hooks, we've moved
+    # this to be its own hook to allow users flexibility to trigger pre-hook before OR after casting
+    # state to match the params' dtype and device.
+    @staticmethod
+    def _cast_state_to_match_params_hook(optimizer, state_dict) -> None:
+        # Validate the state_dict
+        groups = optimizer.param_groups
+        saved_groups = state_dict['param_groups']
+
+        if len(groups) != len(saved_groups):
+            raise ValueError("loaded state dict has a different number of "
+                             "parameter groups")
+        param_lens = (len(g['params']) for g in groups)
+        saved_lens = (len(g['params']) for g in saved_groups)
+        if any(p_len != s_len for p_len, s_len in zip(param_lens, saved_lens)):
+            raise ValueError("loaded state dict contains a parameter group "
+                             "that doesn't match the size of optimizer's group")
+
+        # Update the state_dict with casted values
+        id_map = dict(zip(chain.from_iterable(g['params'] for g in saved_groups),
```

Comment:
If you think this id_map and the associated testing is general, you could have a helper function to get it used both here and in load_state_dict.

Reply:
Reasonable enough.